enforce rref JIT pickling to be in the scope of rpc calls

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34689 enforce rref JIT pickling to be in the scope of rpc calls**

rref JIT pickling is only allowed inside rpc calls. enforcing this by adding a thread local variable isInRpcCall and set it as True when converting rpc requests or responses to message, before calling JIT::pickle(). Inside JIT::pickle(), it allowes to pickle RRef only when the isInRpcCall is true.

Differential Revision: [D20429826](https://our.internmc.facebook.com/intern/diff/D20429826/)

zhaojuanmao committed:
enforce rref JIT pickling to be in the scope of rpc calls

rref JIT pickling is only allowed inside rpc calls. enforcing this by adding a thread local variable isInRpcCall and set it as True when converting rpc requests or responses to message, before calling JIT::pickle(). Inside JIT::pickle(), it allowes to pickle RRef only when the isInRpcCall is true.

Differential Revision: [D20429826](https://our.internmc.facebook.com/intern/diff/D20429826/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_req.cpp:
```diff
@@ -14,10 +14,12 @@ int64_t CleanupAutogradContextReq::getContextId() {
   std::vector<torch::Tensor> tensorTable;
   std::vector<char> payload =
       jit::pickle(at::IValue(context_id_), &tensorTable);
+  rpc::isInRpcCall = false;
```
Is the set/unset needed here? Doesn't seem like we are pickling an RRef.

zhaojuanmao replied:
yes, it ties to jit::pickle, I will rename it. 

when jit pickle the rref, it can check the isInRpcCall flag. I think we should reset it as well, otherwise if user call torch.save() afterward in the same thread, users can pickle rref outside  rpc

rohan-varma commented:
My only concern is that this is somewhat fragile, in that new users would have to ensure that they set/unset this flag properly whenever dealing with pickling in the context of RPC. Although, that may not be too much of an issue as they would likely see an error message and be able to figure it out. Using a scoped guard that sets/unsets this flag may also be a bit cleaner?

zhaojuanmao committed:
Update on "enforce rref JIT pickling to be in the scope of rpc calls"


rref JIT pickling is only allowed inside rpc calls. enforcing this by adding a thread local variable isInRpcCall and set it as True when converting rpc requests or responses to message, before calling JIT::pickle(). Inside JIT::pickle(), it allowes to pickle RRef only when the isInRpcCall is true.

Differential Revision: [D20429826](https://our.internmc.facebook.com/intern/diff/D20429826/)

[ghstack-poisoned]

mrshenli commented:
Any reason for doing this in each individual RPC command's `toMessage()` method instead of in `rpc/python_functions.cpp`?

mrshenli commented on torch/csrc/distributed/rpc/types.h:
```diff
@@ -10,6 +10,16 @@ namespace rpc {
+// Thread local flag to enforce rref JIT pickling to be allowed only
+// in the scope of an rpc call. For other scopes like when model is
+// saved by calling torch.save(), rref is not allowed to be pickled directly.
+extern thread_local bool allowJitRRefPickle;
```
why do we need `extern` keyword here?

mrshenli reviewed (commented):
Question, is there any chance that JIT could use a different thread to perform serialization?

zhaojuanmao commented:
because we could pickle rpc response as well. toMessage covered all the cases

mrshenli commented:
ah, yes, sorry that I forgot that again.

mrshenli commented:
> Question, is there any chance that JIT could use a different thread to perform serialization?

In that case, shall we split `toMessage()` into two methods? Say, define the `toMessage()` in `RpcCommandBase` and another say `toMessageInternal()` (might need a better name) in derived classed. So that we would only need one guard in `toMessage()` method in the base class?

zhaojuanmao commented:
sounds good! looks like so far JIT pickle uses the same thread, I think that make sense, its sync operation

zhaojuanmao committed:
Update on "enforce rref JIT pickling to be in the scope of rpc calls"


rref JIT pickling is only allowed inside rpc calls. enforcing this by adding a thread local variable isInRpcCall and set it as True when converting rpc requests or responses to message, before calling JIT::pickle(). Inside JIT::pickle(), it allowes to pickle RRef only when the isInRpcCall is true.

Differential Revision: [D20429826](https://our.internmc.facebook.com/intern/diff/D20429826/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_command_base.h:
```diff
@@ -11,7 +12,11 @@ class RpcCommandBase {
+    JitRRefPickleGuard jitPickleGuard;
+    return std::move(*this).toMessageInternal();
+  }
+  virtual Message toMessageInternal() && = 0;
```
This should be called `toMessageImpl`, `Internal` kind of indicates some internal details of the class and shouldn't be a public interface method.

mrshenli commented on torch/csrc/jit/serialization/pickler.cpp:
```diff
@@ -136,6 +136,8 @@ void Pickler::pushIValueImpl(const IValue& ivalue) {
   } else if (ivalue.isRRef()) {
 #ifdef USE_DISTRIBUTED
+    TORCH_CHECK(
+        torch::distributed::rpc::allowJitRRefPickle == true, "RRef jit pickling is only allowed inside RPC calls.");
```
will clang-format make this two lines?

pritamdamania87 commented on torch/csrc/distributed/rpc/types.cpp:
```diff
@@ -4,6 +4,9 @@ namespace torch {
 namespace rpc {
 
+// It is declared in torch/csrc/distributed/rpc/types.h
+thread_local bool allowJitRRefPickle = false;
```
Make this static and only define it in the cpp file. Add a separate method in header file `allowJitRRefPickle()` and use that elsewhere in the code.

zhaojuanmao committed:
Update on "enforce rref JIT pickling to be in the scope of rpc calls"


rref JIT pickling is only allowed inside rpc calls. enforcing this by adding a thread local variable isInRpcCall and set it as True when converting rpc requests or responses to message, before calling JIT::pickle(). Inside JIT::pickle(), it allowes to pickle RRef only when the isInRpcCall is true.

Differential Revision: [D20429826](https://our.internmc.facebook.com/intern/diff/D20429826/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/jit/serialization/pickler.cpp:
```diff
@@ -137,7 +137,8 @@ void Pickler::pushIValueImpl(const IValue& ivalue) {
 #ifdef USE_DISTRIBUTED
     TORCH_CHECK(
-        torch::distributed::rpc::allowJitRRefPickle == true, "RRef jit pickling is only allowed inside RPC calls.");
+        torch::distributed::rpc::getAllowJitRRefPickle() == true,
```
nit: don't need `== true`

zhaojuanmao committed:
Update on "enforce rref JIT pickling to be in the scope of rpc calls"


rref JIT pickling is only allowed inside rpc calls. enforcing this by adding a thread local variable isInRpcCall and set it as True when converting rpc requests or responses to message, before calling JIT::pickle(). Inside JIT::pickle(), it allowes to pickle RRef only when the isInRpcCall is true.

Differential Revision: [D20429826](https://our.internmc.facebook.com/intern/diff/D20429826/)

[ghstack-poisoned]
